packages/jss/src/index.d.ts

```diff
@@ -1,49 +1,50 @@
-import {Properties as CSSProperties} from 'csstype'
+import {Properties as CSSProperties, PropertiesHyphen as CSSPropertiesHyphen} from 'csstype'
 
 // Observable support is included as a plugin.  Including it here allows
 // TypeScript users to use Observables, which could be confusing if a user
 // hasn't installed that plugin.
 //
 // TODO: refactor to only include Observable types if plugin is installed.
 export interface MinimalObservable<T> {
-  subscribe(
-    nextOrObserver: ((value: T) => void) | {next: (value: T) => void}
-  ): {unsubscribe: () => void}
+  subscribe(nextOrObserver: ((value: T) => void) | {next: (value: T) => void}): {
+    unsubscribe: () => void
+  }
 }
 
-type Func<P, T, R> = T extends undefined ? ((data: P) => R) : ((data: P & {theme: T}) => R)
+type Func<Data, Theme, Style> = Theme extends undefined
+  ? (data: Data) => Style | null | undefined
+  : (data: Data & {theme: Theme}) => Style | null | undefined
 
-type NormalCssProperties = CSSProperties<string | number>
+type NormalCssProperties = CSSProperties<string | number> & CSSPropertiesHyphen<string | number>
 type NormalCssValues<K> = K extends keyof NormalCssProperties ? NormalCssProperties[K] : JssValue
 
-export type JssStyle<Props = any, Theme = undefined> =
-  | {
-      [K in keyof NormalCssProperties]:
-        | NormalCssValues<K>
-        | JssStyle<Props, Theme>
-        | Func<Props, Theme, NormalCssValues<K> | JssStyle<undefined, undefined> | undefined>
-        | MinimalObservable<NormalCssValues<K> | JssStyle | undefined>
-    }
-  | {
-      [K: string]:
-        | JssValue
-        | JssStyle<Props, Theme>
-        | Func<Props, Theme, JssValue | JssStyle<undefined, undefined> | undefined>
-        | MinimalObservable<JssValue | JssStyle | undefined>
-    }
-
+export type JssStyle<Data = any, Theme = undefined> = {
+  [Key in keyof NormalCssProperties]:
+    | NormalCssValues<Key>
+    | JssStyle<Data, Theme>
+    | Func<Data, Theme, NormalCssValues<Key> | undefined>
+    | MinimalObservable<NormalCssValues<Key> | JssStyle | undefined>
+    | null
+    | undefined
+} & // nesting
+{
+  [Key: `${string}&${string}`]: JssStyle<Data, Theme> | null | undefined
+}
 export type Styles<
   Name extends string | number | symbol = string,
-  Props = unknown,
+  Data = unknown,
```

Comment:
This was supposed to be Data all along, the fact it can be Props is up to user, it could be anything